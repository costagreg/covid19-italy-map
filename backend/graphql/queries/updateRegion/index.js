import { GraphQLList, GraphQLNonNull, GraphQLString } from 'graphql'
import { UpdateRegionType } from '../../types'
import { UpdateRegion } from '../../../database/models'

export const updateRegion = {
  allUpdatesRegion: {
    type: new GraphQLList(UpdateRegionType),
    async resolve() {
      return UpdateRegion.find({})
    },
  },
}
