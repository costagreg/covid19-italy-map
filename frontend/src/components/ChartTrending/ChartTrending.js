import React, { Component } from 'react'
import { scaleBand, scaleLinear, scale } from 'd3-scale'
import { range, max } from 'd3-array'
import { line, curveCatmullRom } from 'd3-shape'
import { format } from 'date-fns'
import './ChartTrending.scss'

export default class ChartTrending extends Component {
  render() {
    const { xAxis, yAxis, chartWidth, chartHeight, marginBottom } = this.props

    const xAxisScale = scaleLinear()
      .domain([0, xAxis.length - 1])
      .range([0, chartWidth])

    const yAxisScale = scaleLinear()
      .domain([0, max(yAxis)])
      .range([chartHeight - marginBottom, marginBottom])

    const chartLine = line()
      .x((d, i) => xAxisScale(i))
      .y((d) => yAxisScale(d))
      .curve(curveCatmullRom.alpha(0.5))

    const xTicks = xAxisScale.ticks(6).map((d) =>
      xAxis[d] ? (
        <g
          transform={`translate(${xAxisScale(d)},${chartHeight -
            marginBottom})`}
        >
          <text className="chartTrending__xTicksText">
            {format(new Date(+xAxis[d]), 'dd/MM')}
          </text>
          <line
            className="chartTrending__xTicksLine"
            x1="0"
            x1="0"
            y1="0"
            y2="5"
          />
        </g>
      ) : null
    )

    return (
      <div className="chartTrending">
        <svg
          width={chartWidth}
          height={chartHeight}
          viewBox={`0 0 ${chartWidth} ${chartHeight}`}
        >
          <g>
            <line
              className="chartTrending__xAxis"
              x1={0}
              x2={chartWidth}
              y1={chartHeight - marginBottom}
              y2={chartHeight - marginBottom}
            />
          </g>
          <g>
            <path className="chartTrending__line" d={chartLine(yAxis)} />
            {yAxis.map((d, i) => (
              <circle cx={xAxisScale(i)} cy={yAxisScale(d)} r={2} />
            ))}
          </g>
          {xTicks}
        </svg>
      </div>
    )
  }
}

ChartTrending.defaultProps = {
  xAxis: [],
  yAxis: [],
}
