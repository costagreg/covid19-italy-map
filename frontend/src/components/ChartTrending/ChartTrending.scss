.chartTrending{
  &__xAxis{
    stroke:#000;
  }
  &__line{
    stroke: steelblue;
    stroke-width: 2px;
    fill: none;
  }
  &__xTicksText{
    font-size: 12px;
    transform:translate(0,15px)
  }

  &__xTicksLine{
    stroke:#000;
    transform:translate(0,0)
  }

}