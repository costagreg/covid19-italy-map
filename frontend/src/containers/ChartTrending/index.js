export { ChartTrending as default } from './ChartTrending'