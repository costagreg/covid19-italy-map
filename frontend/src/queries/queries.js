import gql from 'graphql-tag'

export const FETCH_LATEST_TREND = gql`
  query latestTrendParam($param: String!, $region: String!) {
    latestTrendParam(param: $param, days: 30, region: $region) {
      x
      y
    }
  }
`
export const FETCH_LATEST_UPDATES = gql`
  {
    latestUpdates {
      date
      regions {
        region
        hospitalizedWithSymptoms
        intensiveCare
        totalHospitalized
        homeIsolation
        totalPositive
        totalChangePositive
        newPositive
        dischargedHealed
        totalDeaths
        totalCases
        totalTests
      }
    }
  }
`