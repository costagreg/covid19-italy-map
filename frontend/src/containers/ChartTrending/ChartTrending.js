import React from 'react'
import { Query } from 'react-apollo'
import propTypes from 'prop-types'

import Loader from '../../components/Loader'
import LineChart from '../../components/LineChart'
import { FETCH_LATEST_TREND } from '../../queries'

const ChartTrending = ({ selectedRegion, selectedParam }) => (
  <Query
    query={FETCH_LATEST_TREND}
    variables={{ region: selectedRegion, param: selectedParam }}
  >
    {({ loading, error, data }) => {
      if (loading) return <Loader />
      if (error) return <div>Ops something went wrong</div>

      const { latestTrendParam } = data

      return (
        <div data-testid={`chartTrending-${selectedRegion}`}>
          <LineChart
            xAxis={latestTrendParam.x}
            yAxis={latestTrendParam.y}
            yLabel={selectedParam}
            chartWidth={400}
            chartHeight={200}
            margin={[10, 20, 30, 20]}
          />
        </div>
      )
    }}
  </Query>
)

ChartTrending.propTypes = {
  selectedRegion: propTypes.string.isRequired,
  selectedParam: propTypes.string.isRequired,
}

export default ChartTrending
