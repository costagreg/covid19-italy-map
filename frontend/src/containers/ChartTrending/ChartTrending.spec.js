import React from 'react'
import { render, act } from '@testing-library/react'
import { MockedProvider } from '@apollo/react-testing'
import wait from 'waait'
import ChartTrending from './ChartTrending'
import { FETCH_LATEST_TREND } from '../../queries'

const latestTrendParamSicilia = {
  x: [1, 2],
  y: [10, 20],
  __typename: 'xy',
}

const mocks = [
  {
    request: {
      query: FETCH_LATEST_TREND,
      variables: { region: 'Sicilia', param: 'totalCases' },
    },
    result: {
      data: {
        latestTrendParam: latestTrendParamSicilia,
      },
    },
  },
]

describe('ChartTrending', () => {
  describe('@render', () => {
    it('renders ChartTrending without any error', async () => {
      const { asFragment } = render(
        <MockedProvider mocks={mocks}>
          <ChartTrending selectedRegion="Sicilia" selectedParam="totalCases" />
        </MockedProvider>
      )

      await act(() => wait(0))

      expect(asFragment()).toMatchSnapshot()
    })

    it('shows an error message if data hasn not fetched properly', async () => {
      const { findByText } = render(
        <MockedProvider
          mocks={[
            {
              request: {
                query: FETCH_LATEST_TREND,
                variables: {},
              },
              error: new Error('Something wrong!'),
            },
          ]}
        >
          <ChartTrending selectedRegion="Sicilia" selectedParam="totalCases" />
        </MockedProvider>
      )

      expect(await findByText('Ops something went wrong')).toBeDefined()
    })
  })
})
