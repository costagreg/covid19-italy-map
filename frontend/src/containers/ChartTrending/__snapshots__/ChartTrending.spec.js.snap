// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartTrending @render renders ChartTrending without any error 1`] = `
<DocumentFragment>
  <div
    data-testid="chartTrending-Sicilia"
  >
    <div
      class="lineChart"
    >
      <svg
        height="200"
        viewBox="0 0 400 200"
        width="400"
      >
        <g>
          <line
            class="lineChart__xAxis"
            x1="20"
            x2="380"
            y1="170"
            y2="170"
          />
        </g>
        <g>
          <path
            class="lineChart__line"
            d="M20,90L380,10"
          />
          <g
            transform="translate(20,90)"
          >
            <circle
              class="lineChart__circle"
              r="2"
            />
            <circle
              data-testid="dataPoint"
              fill="transparent"
              r="10"
            />
          </g>
          <g
            transform="translate(380,10)"
          >
            <circle
              class="lineChart__circle"
              r="2"
            />
            <circle
              data-testid="dataPoint"
              fill="transparent"
              r="10"
            />
          </g>
        </g>
        <g
          transform="translate(20,170)"
        >
          <text
            class="lineChart__xTicksText"
          >
            01/01
          </text>
          <line
            class="lineChart__xTicksLine"
            x1="0"
            y1="0"
            y2="5"
          />
        </g>
        <g
          transform="translate(380,170)"
        >
          <text
            class="lineChart__xTicksText"
          >
            01/01
          </text>
          <line
            class="lineChart__xTicksLine"
            x1="0"
            y1="0"
            y2="5"
          />
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
